### PR TITLE
Scala3 ContravariantK macros

### PR DESCRIPTION
**Pull request by pomadchin**
This PR adds a very simple version of the ContravariantK macros. I also noticed that InstanceTests should work x scala versions.

- **ContravariantK**
  - [x] `A => F[_]`
  - [x] `TestAlgebraWithExtraTypeParam`
  - [ ] ...

### CODE REVIEW

File: tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  checkAll("ContravariantK[TestAlgebra]", ContravariantKTests[TestAlgebra].contravariantK[Try, Option, List, Int])
+  checkAll("Serializable ContravariantK[TestAlgebra]", SerializableTests.serializable(ContravariantK[TestAlgebra]))
+
+object autoContravariantKTests:
+  import TestInstances.*
+
+  trait TestAlgebra[F[_]] derives ContravariantK:
+    def sum(xs: F[Int]): Int
+    // ef sumAll(xss: F[Int]*): Int
+    // def foldSpecialized(init: String)(f: (Int, String) => Int): Cokleisli[F, String, Int]
```
Comment (pomadchin) [inline tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala:45]: Eh apparently there are two problems:
1. curried args -- our macro is too simple for that
2. We don't handle construction of Cokleislli; which theoretically is ez via implicit in scope 🤷 but see 1.

File: tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    sa <- Arbitrary.arbitrary[Seq[F[Int]] => Int]
+    fs <- Arbitrary.arbitrary[(String, (Int, String) => Int) => Cokleisli[F, String, Int]]
+  yield new TestAlgebra[F]:
+    def sum(xs: F[Int]) = s(xs)
+    def sumAll(xss: F[Int]*) = sa(xss)
+    def foldSpecialized(init: String)(f: (Int, String) => Int) = fs(init, f)
+  )
+
+  // TODO: Macro should handle it
+  trait TestAlgebraWithExtraTypeParam[A, F[_]] extends TestAlgebra[F] derives ContravariantK:
```
Comment (pomadchin) [inline tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala:66]: the original algebra is `TestAlgebraWithExtraTypeParam[F[_], A]` which causes troubles.

File: tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    def sum(xs: F[Int]) = s(xs)
+    def sumAll(xss: F[Int]*) = sa(xss)
+    def foldSpecialized(init: String)(f: (Int, String) => Int) = fs(init, f)
+  )
+
+  trait TestAlgebraWithExtraTypeParam[F[_], A] extends TestAlgebra[F]:
+    def fold[B](init: B)(f: (B, A) => B): Cokleisli[F, A, B]
+
+  object TestAlgebraWithExtraTypeParam:
+    given contravariantK[T]: ContravariantK[[F[_]] =>> TestAlgebraWithExtraTypeParam[F, T]] = ContravariantK.derived
```
Comment (pomadchin) [inline tests/src/test/scala-3/cats/tagless/tests/autoContravariantKTests.scala:70]: @joroKr21 btw do you know if they fixed it / is there an issue / discussion about it?

**Review by joroKr21 — APPROVED** [review]
Nice work 👍

**Pull request merged**